procdesc.cc

```diff
@@ -853,51 +874,80 @@ TEST_F(PipePdfork, WildcardWait) {
 }
 
 FORK_TEST(Pdfork, Pdkill) {
-  had_signal.clear();
+  clear_had_signals();
   int pd;
+  int pipefds[2];
+  EXPECT_OK(pipe(pipefds));
   pid_t pid = pdfork(&pd, 0);
   EXPECT_OK(pid);
 
   if (pid == 0) {
-    // Child: set a SIGINT handler and sleep.
-    had_signal.clear();
+    // Child: set a SIGINT handler, notify the parent and sleep.
+    close(pipefds[0]);
+    clear_had_signals();
+    last_signal = 0;
     signal(SIGINT, handle_signal);
-    if (verbose) fprintf(stderr, "[%d] child about to sleep(10)\n", getpid_());
-    int left = sleep(10);
-    if (verbose) fprintf(stderr, "[%d] child slept, %d sec left, had[SIGINT]=%d\n",
-                         getpid_(), left, had_signal[SIGINT]);
-    // Expect this sleep to be interrupted by the signal (and so left > 0).
-    exit(left == 0);
+    if (verbose) fprintf(stderr, "[%d] child started\n", getpid_());
+    SEND_INT_MESSAGE(pipefds[1], MSG_CHILD_STARTED);
+    // Child: wait for SIGINT for up to 50 * 100ms seconds.
+    for (volatile int i = 0; i < 50; i++) {
```

Comment:
Why does the `int` need to be `volatile`?

Reply:
That was a workaround when I was debugging why had_signal[SIGINT] was not being read (fixed by using volatile sig_atomic_t instead of std::map<>), so is not longer needed.